secretsclient/client.go

```diff
@@ -629,3 +671,411 @@ func (c *client) ListOrgs() ([]*OrgDetail, error) {
+			Expiry: expiry,
+		})
+	}
+	return tokenDetails, nil
+}
+
+func (c *client) RemoveToken(name string) error {
+	name = url.QueryEscape(name)
+	status, body, err := c.doCall("DELETE", "/api/v0/account/token/"+name, withAuth())
+	if err != nil {
+		return err
+	}
+	if status != http.StatusOK {
+		msg, err := getMessageFromJSON(bytes.NewReader([]byte(body)))
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to decode response body (status code: %d)", status))
+		}
+		return fmt.Errorf("failed to delete token: %s", msg)
+	}
+	return nil
+}
+
+func (c *client) WhoAmI() (string, string, error) {
+	status, body, err := c.doCall("GET", "/api/v0/account/ping", withAuth())
+	if err != nil {
+		return "", "", err
+	}
+	if status != http.StatusOK {
+		msg, err := getMessageFromJSON(bytes.NewReader([]byte(body)))
+		if err != nil {
+			return "", "", errors.Wrap(err, fmt.Sprintf("failed to decode response body (status code: %d)", status))
+		}
+		return "", "", fmt.Errorf("failed to authenticate: %s", msg)
+	}
+
+	var pingResponse api.PingResponse
+	err = jsonpb.Unmarshal(bytes.NewReader([]byte(body)), &pingResponse)
+	if err != nil {
+		return "", "", errors.Wrap(err, "failed to unmarshal ping response")
+	}
+
+	authType := "ssh"
+	if c.password != "" {
+		authType = "password"
+	} else if c.authToken != "" {
+		authType = "token"
+	}
+
+	return pingResponse.Email, authType, nil
+}
+
+func (c *client) getAuthTokenPath(create bool) (string, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get home dir")
+	}
+
+	confDirPath := filepath.Join(homeDir, ".earthly")
+	if !fileutils.DirExists(confDirPath) && create {
+		err := os.MkdirAll(confDirPath, 0755)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to create run directory %s", confDirPath)
+		}
+	}
+
+	tokenPath := filepath.Join(confDirPath, "auth.token")
+	return tokenPath, nil
+}
+
+// loads ~/.earthly/auth.token
+// which is formatted as
+// <email> <type> ...
+func (c *client) loadAuthToken() error {
+	tokenPath, err := c.getAuthTokenPath(false)
+	if err != nil {
+		return err
+	}
+	if !fileutils.FileExists(tokenPath) {
+		return nil
+	}
+	data, err := ioutil.ReadFile(tokenPath)
+	if err != nil {
+		return errors.Wrap(err, "failed to read file")
+	}
+	parts := strings.SplitN(string(data), " ", 3)
+	if len(parts) != 3 {
+		return nil
+	}
+	c.email = parts[0]
+	authType := parts[1]
+	authData := parts[2]
+	switch authType {
+	case "password":
+		passwordBytes, err := base64.StdEncoding.DecodeString(authData)
+		if err != nil {
+			return errors.Wrap(err, "base64 decode failed")
+		}
+		c.password = string(passwordBytes)
+	case "ssh-rsa":
+		c.sshKeyBlob, err = base64.StdEncoding.DecodeString(authData)
+		if err != nil {
+			return errors.Wrap(err, "base64 decode failed")
+		}
+	case "token":
+		c.authToken = authData
+	default:
+		c.warnFunc("unable to handle cached auth type %s", authType)
+		break
+	}
+	return nil
+}
```

Comment:
We should encrypt password / token so they are not stored in the clear on the disk. In any case, this can be a future thing.

(Docker uses this credentialsStore: https://docs.docker.com/engine/reference/commandline/login/#credentials-store which attempts to use common tools / key chains)

For now, we can recommend users to use SSH key + passphrase if they want to avoid this.

Reply:
As discussed in our chat, we'll do our best to prompt the user to add an ssh key wherever possible, and print a warning in red whenever we revert to caching the password on disk. In a future PR we can also extend the server to return some sort of signed auth token similar to the credentials-store.